From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: commandblockguy <commandblockguy1@gmail.com>
Date: Fri, 14 Aug 2020 14:44:14 -0500
Subject: [PATCH] Prevent headless pistons from being created

Prevent headless pistons from being created by explosions or tree/mushroom growth.

diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index d86474a1a97597c14ea8a28444e2a5d82fcda2a7..e0422471776f34b8e3c753d855aa06f00cc049ec 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -436,6 +436,12 @@ public class PaperConfig {
         set("settings.unsupported-settings.allow-tnt-duplication", null);
     }
 
+    public static boolean allowHeadlessPistons;
+    private static void allowHeadlessPistons() {
+        config.set("settings.unsupported-settings.allow-headless-pistons-readme", "This setting controls if players should be able to create headless pistons.");
+        allowHeadlessPistons = getBoolean("settings.unsupported-settings.allow-headless-pistons", false);
+    }
+
     public static int playerAutoSaveRate = -1;
     public static int maxPlayerAutoSavePerTick = 10;
     private static void playerAutoSaveRate() {
diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index 72de544c0782e4daff222a0ca04e5bc870cf148c..5ca4d03215c979a4d122a32178a1859b4d9c3334 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -135,6 +135,15 @@ public class Explosion {
 
                             if (f > 0.0F && this.l.a(this, this.world, blockposition, iblockdata, f) && blockposition.getY() < 256 && blockposition.getY() >= 0) { // CraftBukkit - don't wrap explosions
                                 set.add(blockposition);
+                                // Paper start - prevent headless pistons from forming
+                                if (!com.destroystokyo.paper.PaperConfig.allowHeadlessPistons && iblockdata.getBlock() == Blocks.MOVING_PISTON) {
+                                    TileEntity extension = this.world.getTileEntity(blockposition);
+                                    if (extension instanceof TileEntityPiston && ((TileEntityPiston) extension).isHead()) {
+                                       EnumDirection direction = iblockdata.get(BlockPistonExtension.FACING);
+                                       set.add(blockposition.shift(direction.opposite()));
+                                    }
+                                }
+                                // Paper end
                             }
 
                             d4 += d0 * 0.30000001192092896D;
diff --git a/src/main/java/net/minecraft/server/TileEntityPiston.java b/src/main/java/net/minecraft/server/TileEntityPiston.java
index 7d86f0fe55063f6875db9c6f99f4f72ed4144536..b0d52d7405ca24938ca753897f89e6fad575d04d 100644
--- a/src/main/java/net/minecraft/server/TileEntityPiston.java
+++ b/src/main/java/net/minecraft/server/TileEntityPiston.java
@@ -42,6 +42,8 @@ public class TileEntityPiston extends TileEntity implements ITickable {
         return this.b;
     }
 
+    public final boolean isHead() { return this.h(); } // Paper - OBFHELPER
+
     public boolean h() {
         return this.g;
     }
